File: c++/Decompressor.cpp
#include "Decompressor.h"

unsigned char Decompressor::get_char(const unsigned long long int &low, const unsigned long long int &value, const unsigned long long int &high){
    std::vector<unsigned int> chars = model.get_keys();
    unsigned long long int l = high - low;
    for (auto c: chars){
        Interval p = model.get_interval(c);
        unsigned long long int high_tmp = low + (l * p.upper)/model.get_count();
        unsigned long long int low_tmp = low + (l * p.lower)/model.get_count();
        if(low_tmp <= value && value < high_tmp)
            return c;
    }
    return chars[0];
}

void Decompressor::decompress(){
    struct SP{
        unsigned long long int PRECISION;
        unsigned long long int HALF;
        unsigned long long int MAX;
        unsigned long long int QUARTER;
        unsigned long long int TQUARTER;
    };
    int p = 28;
    SP SPECS;
    SPECS.PRECISION = p;
    SPECS.MAX = pow(2,p);
    SPECS.HALF = pow(2,p-1);
    SPECS.QUARTER = pow(2,p-2);
    SPECS.TQUARTER = 3*pow(2,p-2);
    unsigned long long int high = SPECS.MAX;
    unsigned long long int low = 0;
    unsigned long long int value = 0;
    bool bit;
    // float f = 0xFFFFFFFFU;
    for(int i=0; i<SPECS.PRECISION; i++){
        read_stream >> bit;
        value<<=1;
        value += bit;
    }
    //std::cout << value << std::endl;
    for(int j=0; j<model.get_count()-1; j++){
        unsigned long long int l = high - low;
        unsigned int c = get_char(low,value,high);
        Interval p = model.get_interval(c);
        write_stream << (unsigned char)c;
        high = low + (l * p.upper)/model.get_count();
        low = low + (l * p.lower)/model.get_count();
        //std::cout << std::endl<< (unsigned char)c <<" - "<< j << std::endl;
        for( ; ; ){
            //std::cout << low << ", " << value << ", " << high << std::endl;
            if(high < SPECS.HALF){
                low <<=1;
                high <<=1;
                value <<=1;
                read_stream >> bit;
                value += bit;
            }
            else if(low >= SPECS.HALF){
                low = 2*(low-SPECS.HALF);
                high = 2*(high-SPECS.HALF);
                value = 2*(value-SPECS.HALF);
                read_stream >> bit;
                value += bit;
            }
            else if(low>SPECS.QUARTER && high<SPECS.TQUARTER){
                low = 2*(low-SPECS.QUARTER);
                high = 2*(high-SPECS.QUARTER);
                value = 2*(value-SPECS.QUARTER);
                read_stream >> bit;
                value += bit;
            }
            else{
                break;
            }
        }
    }
    write_stream << "\n";
    write_stream.close();
}

File: python/Decompressor.py
import os
from Read_bit import Read_bit
from specs import SPECS

class Decompressor:
    def __init__(self,namefile,model):
        self.write_file = open(os.path.splitext(namefile)[0]+'.decode', "wb")
        self.read_file = Read_bit(namefile)
        self.model = model

    def decompress(self):
        high = SPECS["MAX"]
        low = 0
        value = 0
        for i in range(SPECS["PRECISION"]):
            value<<=1
            value+=self.read_file.read()
        for ii in range(self.model.count):
            l = high - low
            # high_tmp = high
            # low_tmp = low
            # for k in self.model.keys:
            #     (lower,upper) = self.model.c[k]
            #     high_tmp = low + (l * upper)/self.model.count
            #     low_tmp = low + (l * lower)/self.model.count
            #     if(low_tmp <= value and value < high_tmp):
            #         s = k
            s = self.get_char(low,value,high)
            print chr(s)
            self.write_file.write(chr(s))
            p = self.model.c[s]
            high = low + (l * p[1])/self.model.count
            low = low + (l * p[0])/self.model.count
            check = True
            while(check):
                print low,high
                if(high < SPECS["HALF"]):
                    high <<= 1
                    low <<= 1
                    value <<= 1
                    value += self.read_file.read()
                elif(low > SPECS["HALF"]):
                    high = (high - SPECS["HALF"])<<1
                    value = (value - SPECS["HALF"])<<1
                    low = (low - SPECS["HALF"])<<1
                    value += self.read_file.read()
                elif(low>SPECS["QUARTER"] and high<SPECS["TQUARTER"]):
                    high = (high - SPECS["QUARTER"])<<1
                    low = (low - SPECS["QUARTER"])<<1
                    value = (value - SPECS["QUARTER"])<<1
                    value += self.read_file.read()
                else: check = False
        self.read_file.close()
        self.write_file.close()

    def get_char(self,low,value,high):
        l = high - low
        for k in self.model.keys:
            (lower,upper) = self.model.c[k]
            high_tmp = low + (l * upper)/self.model.count
            low_tmp = low + (l * lower)/self.model.count
            if(low_tmp <= value and value < high_tmp):
                return k
        return k
